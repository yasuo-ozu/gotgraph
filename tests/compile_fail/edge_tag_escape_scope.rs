use gotgraph::prelude::*;

fn main() {
    let mut graph: VecGraph<i32, &str> = VecGraph::default();
    
    // Try to move edge tags outside of scope - should fail
    let escaped_edge = graph.scope_mut(|mut ctx| {
        let n0 = ctx.add_node(1);
        let n1 = ctx.add_node(2);
        let edge = ctx.add_edge("test", n0, n1);
        edge // ERROR: edge tag cannot escape the scope
    });
    
    // Try to use the escaped edge tag - should fail
    println!("{:?}", escaped_edge);
}