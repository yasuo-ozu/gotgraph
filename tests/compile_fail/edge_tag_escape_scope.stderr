error: lifetime may not live long enough
  --> tests/compile_fail/edge_tag_escape_scope.rs:11:9
   |
 7 |     let escaped_edge = graph.scope_mut(|mut ctx| {
   |                                         -------- return type of closure is EdgeTag<'2, EdgeIx>
   |                                         |
   |                                         has type `gotgraph::graph::Context<'1, &mut gotgraph::vec_graph::VecGraph<i32, &str>>`
...
11 |         edge // ERROR: edge tag cannot escape the scope
   |         ^^^^ returning this value requires that `'1` must outlive `'2`
   |
   = note: requirement occurs because of the type `EdgeTag<'_, EdgeIx>`, which makes the generic argument `'_` invariant
   = note: the struct `EdgeTag<'scope, I>` is invariant over the parameter `'scope`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
